src/optimizer/rule/PushFilterDownAggregateRule.cpp

```diff
@@ -0,0 +1,127 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/PushFilterDownAggregateRule.h"
+
+#include "optimizer/OptContext.h"
+#include "optimizer/OptGroup.h"
+#include "planner/PlanNode.h"
+#include "planner/Query.h"
+#include "util/ExpressionUtils.h"
+#include "visitor/ExtractFilterExprVisitor.h"
+#include "visitor/RewriteVisitor.h"
+
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> PushFilterDownAggregateRule::kInstance =
+    std::unique_ptr<PushFilterDownAggregateRule>(new PushFilterDownAggregateRule());
+
+PushFilterDownAggregateRule::PushFilterDownAggregateRule() {
+    RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& PushFilterDownAggregateRule::pattern() const {
+    static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kFilter,
+                                             {Pattern::create(graph::PlanNode::Kind::kAggregate)});
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> PushFilterDownAggregateRule::transform(
+    OptContext* octx,
+    const MatchedResult& matched) const {
+    auto qctx = octx->qctx();
+    auto* filterGroupNode = matched.node;
+    auto* oldFilterNode = filterGroupNode->node();
+    auto deps = matched.dependencies;
+    DCHECK_EQ(deps.size(), 1);
+    auto aggGroupNode = deps.front().node;
+    auto* oldAggNode = aggGroupNode->node();
+    DCHECK(oldFilterNode->kind() == PlanNode::Kind::kFilter);
+    DCHECK(oldAggNode->kind() == PlanNode::Kind::kAggregate);
+    auto* newFilterNode =
+        const_cast<graph::Filter*>(static_cast<const graph::Filter*>(oldFilterNode))->clone(qctx);
```

Comment:
Need not to cast the pointer to nonconst Filter since clone is a const function

Reply:
ok